components/turnToAngle.py

```diff
@@ -1,101 +1,97 @@
+    def setAngle(self, angle):
+        """Sets the desired turn angle"""
+        self.turnAngle = angle
 
-    @feedback
-    def outputDisplay(self):
-        return self.nextOutput
+    def start(self):
+        self.starting = True
 
-    @feedback
-    def nextHeadingDisplay(self):
-        return self.nextHeading
+    @state(first = True)
+    def idling(self):
+        """Stays in this state until started"""
+        if self.turnAngle != 0:
+            self.next_state("turn")
+        else:
+            log.error("Must have an angle to turn to")
+            self.next_state("idling")
+
+
+    def calcHeading(self):
+        """Calculates how far away from the desired angle the bot is"""
+        self.running = True
+        self.starting = False
+        self.nextHeading = self.initialHeading + self.turnAngle
 
-    @feedback
-    def setSpeedDisplay(self):
-        return self.setSpeed
-    
-    def stop(self):
-        self.nextOutput = 0
-        self.PIDController.reset()
-        
         if self.nextHeading > 360:
             self.nextHeading -= 360
-        
-        self.isRunning = False
-        self.initialHeading = self.heading
-        self.setSpeed = True
+        elif self.nextHeading < 0:
+            self.nextHeading += 360
 
+        self.change = self.nextHeading - self.navx.getFusedHeading()
+        if self.change > 180:
+            self.change -= 360
+        elif self.change < -180:
+            self.change += 360
 
-    def execute(self):
-        self.output()
-        self.PIDController = controller.PIDController(Kp= self.P, Ki= self.I, Kd= self.D, period = self.time)
-        self.heading = self.navx.getFusedHeading()
+    def setSpeedFunc(self):
+        """Determines the speed based off of the distance from the desired angle"""
+        self.calcHeading()
+        if abs(self.change) > 90:
+            self.speed = self.dumbSpeed * self.farMultiplier
+        elif abs(self.change) <= 90 and abs(self.change) > 20:
+            self.speed = self.dumbSpeed * self.midMultiplier
+        elif abs(self.change) <= 20:
+            self.speed = self.dumbSpeed * self.closeMultiplier
+
+    @state
+    def turn(self):
+        """Turns the robot based off of the speed determined in setSpeedFunc"""
+        self.setSpeedFunc()
+        if self.change > 0:
+            self.driveTrain.setTank(-1 * self.speed, self.speed)
+        else:
+            self.driveTrain.setTank(self.speed, -1 * self.speed)
+        self.next_state("turn")
+
+        """Stops the automatic turning if the bot is within the tolerance of the desired angle"""
+        if self.navx.getFusedHeading() <= self.nextHeading + self.tolerance and self.navx.getFusedHeading() >= self.nextHeading - self.tolerance:
+            self.driveTrain.setTank(0, 0)
+            self.stop()
+
+    def stop(self):
+        self.running = False
+        self.starting = False
+        self.initialHeading = self.navx.getFusedHeading()
```

Comment:
Why get the fused heading here? Seems like just need to get it when a new angle is commanded? Might not be the same when we start moving again.

Reply:
Shoudn't this set the angle goal to 0 to stop turning?